File: Striver-A2Z/13_Binary_Tree/2_Medium/7_Boundary_Traversal.cpp

#include <bits/stdc++.h>

using namespace std;

struct node
{
  int data;
  struct node *left, *right;
};

bool isLeaf(node *root)
{
  return !root->left && !root->right;
}

void addLeftBoundary(node *root, vector<int> &res)
{
  node *cur = root->left;
  while (cur)
  {
    if (!isLeaf(cur))
      res.push_back(cur->data);
    if (cur->left)
      cur = cur->left;
    else
      cur = cur->right;
  }
}
void addRightBoundary(node *root, vector<int> &res)
{
  node *cur = root->right;
  vector<int> tmp;
  while (cur)
  {
    if (!isLeaf(cur))
      tmp.push_back(cur->data);
    if (cur->right)
      cur = cur->right;
    else
      cur = cur->left;
  }
  for (int i = tmp.size() - 1; i >= 0; --i)
  {
    res.push_back(tmp[i]);
  }
}

void addLeaves(node *root, vector<int> &res)
{
  if (isLeaf(root))
  {
    res.push_back(root->data);
    return;
  }
  if (root->left)
    addLeaves(root->left, res);
  if (root->right)
    addLeaves(root->right, res);
}

vector<int> printBoundary(node *root)
{
  vector<int> res;
  if (!root)
    return res;

  if (!isLeaf(root))
    res.push_back(root->data);

  addLeftBoundary(root, res);

  // add leaf nodes
  addLeaves(root, res);

  addRightBoundary(root, res);
  return res;
}

struct node *newNode(int data)
{
  struct node *node = (struct node *)malloc(sizeof(struct node));
  node->data = data;
  node->left = NULL;
  node->right = NULL;

  return (node);
}

int main()
{
  struct node *root = newNode(4);
  root->left = newNode(10);

  root->left->right = newNode(5);
  root->left->left = newNode(5);

  root->left->left->right = newNode(6);
  root->left->right->left = newNode(7);

  // root->left->left->right->left = newNode(5);
  // root->left->left->right->right = newNode(6);

  // root->right->right->left = newNode(9);
  // root->right->right->left->left = newNode(10);
  // root->right->right->left->right = newNode(11);

  vector<int> boundaryTraversal;
  boundaryTraversal = printBoundary(root);

  cout << "The Boundary Traversal is : ";
  for (int i = 0; i < boundaryTraversal.size(); i++)
  {
    cout << boundaryTraversal[i] << " ";
  }
  return 0;
}

// 4 10 N 5 5 N 6 7 N 8 8 N 8 11 N 3 4 N 1 3 N 8 6 N 11 11 N 5 8.
// [4, 10, null, 5, 5, null, 6, 7, null, 8 ,8,null, 8, 11, null, 3, 4, null, 1, 3, null, 8, 6, null, 11, 11, null, 5, 8]

File: Striver-A2Z/13_Binary_Tree/2_Medium/7_Boundary_Traversal.py
""" 
Given a Binary Tree, find its Boundary Traversal. The traversal should be in the following order: 

Left boundary nodes: defined as the path from the root to the left-most node ie- the leaf node you could reach when you always travel preferring the left subtree over the right subtree. 
Leaf nodes: All the leaf nodes except for the ones that are part of left or right boundary.
Reverse right boundary nodes: defined as the path from the right-most node to the root. The right-most node is the leaf node you could reach when you always travel preferring the right subtree over the left subtree. Exclude the root from this as it was already included in the traversal of left boundary nodes.
Note: If the root doesn't have a left subtree or right subtree, then the root itself is the left or right boundary. 

"""


class Solution:
    def printBoundaryView(self, root):

        if root.left == None and root.right == None:
            return [root.data]

        left_bound = []
        temp = root.left

        if temp:
            while temp:
                left_bound.append(temp.data)

                if temp.left:
                    temp = temp.left
                else:
                    temp = temp.right

        leaf_node = []

        def leaf(temp):
            if temp:
                if not temp.left and not temp.right:
                    leaf_node.append(temp.data)

                leaf(temp.left)
                leaf(temp.right)
        leaf(root)

        # print(leaf_node)

        if len(leaf_node) > 0 and len(left_bound) > 0:
            leaf_node.pop(0)

        right_bound = []
        temp = root.right

        if temp:
            while temp:
                right_bound.append(temp.data)
                if temp.right:
                    temp = temp.right
                else:
                    temp = temp.left

        if len(right_bound) > 0:
            right_bound.pop()

        right_bound.reverse()

        res = [root.data]
        res.extend(left_bound)
        res.extend(leaf_node)
        res.extend(right_bound)

        return res
